File: android/app/src/main/java/com/lttprandomizer/AlttprApiClient.kt
package com.lttprandomizer

import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

object AlttprApiClient {
    internal val json = Json { encodeDefaults = true; ignoreUnknownKeys = true }

    internal val http = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .addInterceptor { chain ->
            chain.proceed(
                chain.request().newBuilder()
                    .header("User-Agent", "LTTPRandomizerGenerator-Android/0.1")
                    .build()
            )
        }
        .build()

    private val JSON_MEDIA = "application/json; charset=utf-8".toMediaType()
    private const val BASE = "https://alttpr.com"

    data class SeedResult(
        val hash: String,
        val permalink: String,
        val bpsBytes: ByteArray,
        val dictPatches: List<Map<String, List<Int>>>,
        val sizeMb: Int,
    )

    @kotlinx.serialization.Serializable
    private data class HashInfo(
        @kotlinx.serialization.SerialName("bpsLocation") val bpsLocation: String = "",
        @kotlinx.serialization.SerialName("md5")         val md5: String = "",
    )

    /**
     * Generates a seed and returns [SeedResult].
     * Throws [IOException] or [IllegalStateException] on failure.
     * Call from a coroutine (not the main thread).
     *
     * Two-step flow (mirrors pyz3r):
     *   1. POST /api/randomizer  → seed hash + dict patches
     *   2. GET  /api/h/{hash}    → bpsLocation for the base BPS patch
     *   3. GET  bpsLocation      → download BPS (English translation + engine)
     */
    fun generate(settings: RandomizerSettings, onProgress: (String) -> Unit): SeedResult {
        onProgress("Contacting alttpr.com…")

        val body = json.encodeToString(settings).toRequestBody(JSON_MEDIA)
        val apiResponse: SeedApiResponse = http.newCall(
            Request.Builder().url("$BASE/api/randomizer").post(body).build()
        ).execute().use { resp ->
            if (!resp.isSuccessful)
                throw IOException("API error ${resp.code}: ${resp.message}")
            val raw = resp.body?.string() ?: throw IOException("Empty API response")
            json.decodeFromString(raw)
        }

        onProgress("Fetching patch metadata…")
        val hashInfo: HashInfo = http.newCall(
            Request.Builder().url("$BASE/api/h/${apiResponse.hash}").build()
        ).execute().use { resp ->
            if (!resp.isSuccessful)
                throw IOException("Failed to fetch patch metadata: ${resp.code}")
            val raw = resp.body?.string() ?: throw IOException("Empty metadata response")
            json.decodeFromString(raw)
        }

        onProgress("Downloading base patch…")
        val bpsBytes = if (hashInfo.bpsLocation.isNotBlank()) {
            val bpsUrl = if (hashInfo.bpsLocation.startsWith("http")) hashInfo.bpsLocation
                         else "$BASE${hashInfo.bpsLocation}"
            http.newCall(Request.Builder().url(bpsUrl).build())
                .execute().use { resp ->
                    if (!resp.isSuccessful)
                        throw IOException("Failed to download BPS patch: ${resp.code}")
                    resp.body?.bytes() ?: throw IOException("Empty BPS response")
                }
        } else ByteArray(0)

        return SeedResult(
            hash        = apiResponse.hash,
            permalink   = "$BASE/h/${apiResponse.hash}",
            bpsBytes    = bpsBytes,
            dictPatches = apiResponse.patch,
            sizeMb      = apiResponse.size,
        )
    }
}

File: windows/Services/AlttprApiClient.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;
using LTTPRandomizerGenerator.Models;

namespace LTTPRandomizerGenerator.Services
{
    /// <summary>
    /// Communicates with the alttpr.com API to generate a randomizer seed
    /// and retrieve the patch data needed to produce the output ROM.
    /// </summary>
    public static class AlttprApiClient
    {
        private static readonly HttpClient Http = new()
        {
            BaseAddress = new Uri("https://alttpr.com"),
            Timeout = TimeSpan.FromSeconds(60),
        };

        static AlttprApiClient()
        {
            Http.DefaultRequestHeaders.Add("User-Agent", "LTTPRandomizerGenerator/0.1");
        }

        /// <summary>
        /// Posts settings to the API and returns the seed result.
        /// Follows pyz3r's two-step flow: POST /api/randomizer for seed data,
        /// then GET /api/h/{hash} to obtain bpsLocation for the base BPS patch.
        /// The BPS patch contains the English translation and randomizer engine;
        /// without it, dict patches apply to the raw Japanese ROM (garbled text).
        /// </summary>
        public static async Task<SeedResult?> GenerateAsync(
            RandomizerSettings settings,
            IProgress<string>? progress = null,
            CancellationToken ct = default)
        {
            // Step 1 — generate the seed
            progress?.Report("Contacting alttpr.com...");
            HttpResponseMessage response;
            try
            {
                response = await Http.PostAsJsonAsync("/api/randomizer", settings, ct);
            }
            catch (HttpRequestException ex)
            {
                throw new InvalidOperationException($"Network error: {ex.Message}", ex);
            }
            catch (TaskCanceledException)
            {
                throw new InvalidOperationException("Request timed out. Check your internet connection.");
            }

            if (!response.IsSuccessStatusCode)
            {
                string body = await response.Content.ReadAsStringAsync(ct);
                throw new InvalidOperationException(
                    $"API returned {(int)response.StatusCode}: {response.ReasonPhrase}\n{body}");
            }

            progress?.Report("Parsing seed data...");
            var apiResponse = await response.Content.ReadFromJsonAsync<ApiResponse>(ct);
            if (apiResponse is null)
                throw new InvalidOperationException("API returned empty response.");

            // Step 2 — fetch bpsLocation from /api/h/{hash} (mirrors pyz3r's get_patch_base())
            progress?.Report("Fetching patch metadata...");
            HashInfo? hashInfo;
            try
            {
                hashInfo = await Http.GetFromJsonAsync<HashInfo>($"/api/h/{apiResponse.Hash}", ct);
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException($"Failed to fetch patch metadata: {ex.Message}", ex);
            }

            // Step 3 — download the BPS base patch (English translation + randomizer engine)
            byte[] bpsBytes = Array.Empty<byte>();
            string? bpsLocation = hashInfo?.BpsLocation;
            if (!string.IsNullOrWhiteSpace(bpsLocation))
            {
                progress?.Report("Downloading base patch...");
                try
                {
                    bpsBytes = await Http.GetByteArrayAsync(bpsLocation, ct);
                }
                catch (Exception ex)
                {
                    throw new InvalidOperationException($"Failed to download base patch: {ex.Message}", ex);
                }
            }

            return new SeedResult
            {
                Hash = apiResponse.Hash,
                Permalink = $"https://alttpr.com/h/{apiResponse.Hash}",
                BpsPatchBytes = bpsBytes,
                DictionaryPatches = apiResponse.Patch,
                RomSizeMb = apiResponse.Size,
            };
        }

        // ── Internal deserialization types ────────────────────────────────────

        private class ApiResponse
        {
            [JsonPropertyName("hash")]
            public string Hash { get; set; } = string.Empty;

            [JsonPropertyName("patch")]
            public List<Dictionary<string, List<int>>> Patch { get; set; } = new();

            [JsonPropertyName("size")]
            public int Size { get; set; } = 2;
        }

        private class HashInfo
        {
            [JsonPropertyName("bpsLocation")]
            public string BpsLocation { get; set; } = string.Empty;

            [JsonPropertyName("md5")]
            public string Md5 { get; set; } = string.Empty;
        }
    }

    public class SeedResult
    {
        public string Hash { get; set; } = string.Empty;
        public string Permalink { get; set; } = string.Empty;
        public byte[] BpsPatchBytes { get; set; } = Array.Empty<byte>();
        public List<Dictionary<string, List<int>>> DictionaryPatches { get; set; } = new();
        public int RomSizeMb { get; set; } = 2;
    }
}
